crates/bevy_ecs/src/world/command_queue.rs

```diff
@@ -120,7 +120,21 @@ impl CommandQueue {
         // flush the previously queued entities
         world.flush_entities();
 
-        self.apply_or_drop_queued(Some(world));
+        // If the world's command queue is empty replace it with our queue
+        if world.command_queue.is_empty() {
+            std::mem::swap(self, &mut world.command_queue);
+        }
+
+        // Apply any commands in the world's internal queue first.
+        while !world.command_queue.is_empty() {
+            std::mem::swap(self, &mut world.command_queue);
+            self.apply_or_drop_queued(Some(world));
+        }
+
+        // Return the larger buffer to the world to be re-used
+        if self.bytes.capacity() > world.command_queue.bytes.capacity() {
+            std::mem::swap(self, &mut world.command_queue);
+        }
```

Comment:
Wondering if this juggling is necessary -- it makes the control flow a bit hard to follow and I'm still not 100% sure if every buffer will get put back in the right place.

Could we double-buffer the world's command queues? There could be a separate queue stored stored in the world as a private field, which we'd swap with the main queue when applying commands to reuse allocations. That should also prevent the capacity of unrelated system command queues from being inflated by commands added to the world's queue

Reply:
Yes, double buffering is a suitable alternative and I can swap to that instead.

Reply:
I would prefer that. It seems simpler and more memory-efficient. It might end up making the code in this loop longer but I imagine it would compile to similar code after being unrolled

Reply:
Fixed.

I wouldn't say it alleviates the juggling since you can't apply the additional buffer while it's still in the world, but it is better for ensuring buffers don't grow unnecessarily.